runtime/near-vm-runner/src/logic/logic.rs

```diff
@@ -3187,6 +3216,45 @@ bls12381_p2_decompress_base + bls12381_p2_decompress_element * num_elements`
         Ok(self.context.promise_results.len() as _)
     }
 
+    /// A method to check how much memory [`Self::promise_result()`] would return.
+    ///
+    /// If the current function is invoked by a callback, we can access the length of execution
+    /// results of the promises that caused the callback. It can be used to prevent out-of-gas
+    /// failures when reading too long execution result via [`Self::promise_result()`].
+    ///
+    /// # Returns
+    ///
+    /// * If promise result is complete and successful returns the length in bytes of the promise result;
+    /// * If promise result is not complete or failed returns `0`;
+    ///
+    /// # Errors
+    ///
+    /// * If `result_idx` does not correspond to an existing result returns [`HostError::InvalidPromiseResultIndex`];
+    /// * If called as view function returns [`HostError::ProhibitedInView`].
+    ///
+    /// # Cost
+    ///
+    /// `base` - the base cost for a simple host function call
+    pub fn promise_result_length(&mut self, result_idx: u64) -> Result<u64> {
```

Comment:
What is the use-case for this host function, really? The description here says its about avoiding gas exhaustion, but what else would a contract do if the result is too long?

I find it somewhat not satisfying that this is a mechanism to avoid the problem of "data too long" for just one host function out of many that interact with registers. I feel it might be worth exploring if there is an opportunity for a size limiting mechanism that could be applied generally across everything dealing with registers.